client.go

```diff
@@ -474,27 +477,25 @@ func (c *Client) Do(req *Request) (*http.Response, error) {
 		// Check if we should continue with retries.
 		checkOK, checkErr := c.CheckRetry(req.Context(), resp, err)
 
-		if logger != nil {
-			if err != nil {
+		if err != nil {
+			switch v := logger.(type) {
+			case Logger:
+				v.Printf("[ERR] %s %s request failed: %v", req.Method, req.URL, err)
+			case hclog.Logger:
+				v.Error("request failed", "error", err, "method", req.Method, "url", req.URL)
+			}
+		} else {
+			// Call this here to maintain the behavior of logging all requests,
+			// even if CheckRetry signals to stop.
+			if c.ResponseLogHook != nil {
+				// Call the response logger function if provided.
 				switch v := logger.(type) {
 				case Logger:
-					v.Printf("[ERR] %s %s request failed: %v", req.Method, req.URL, err)
+					c.ResponseLogHook(v, resp)
 				case hclog.Logger:
-					v.Error("request failed", "error", err, "method", req.Method, "url", req.URL)
-				}
-			} else {
-				// Call this here to maintain the behavior of logging all requests,
-				// even if CheckRetry signals to stop.
-				if c.ResponseLogHook != nil {
-					// Call the response logger function if provided.
-					switch v := logger.(type) {
-					case Logger:
-						c.ResponseLogHook(v, resp)
-					case hclog.Logger:
-						c.ResponseLogHook(hookLogger{v}, resp)
-					default:
-						c.ResponseLogHook(nil, resp)
-					}
+					c.ResponseLogHook(hookLogger{v}, resp)
+				case nil:
```

Comment:
Should use default instead of nil here in the offhand case that someone passes a strange logger.

Reply:
Changed